### DISCUSSION TITLE
HTOdemux with large hashtag imbalance

### DISCUSSION THREAD

**Discussion by erosix** [satijalab/seurat#5110] [Q&A]
Hello,
I have a dataset of 10xGenomics 5' Gene expression data + hashtags. 
In details there are 3 hashtags. Example of hashtag count table:
```
HASH1   16  21   12   13 777   14    2    8    8  3    9
HASH2  1376 167 1061 2016  66 1721 3148 2995 1999 55 1556
HASH3    6   8    8    3   4    3    7    4    3 32    3

```
When I run HTOdemux, the cutoff threshold is very imbalanced, examples:
Sample1:
```
Cutoff for HASH1 : 26 reads
Cutoff for HASH2  : 3032 reads
Cutoff for HASH3  : 15 reads
```
Sample 2: 
```
Cutoff for HASH1.1 : 21 reads
Cutoff for HASH2 : 582 reads
Cutoff for HASH3 : 8 reads
```

The cells with hashtag "HASH2" are much more than the ones with hashtag 1 and 3 (which is what we expect). They should be approx. 10 times more, may this explain the strong imbalance in the hashtag cutoff?

The problem is that because of these high cutoffs for HASH2, a large proportion of cells are actually assigned as "negative", example:
Sample1:
```
HTO.global:
Doublet Negative  Singlet 
     522     3191     2161 
```

```
HTO:
HASH2 : 1126
HASH1: 529
HASH3: 506 
Negative: 3191 
```

I wonder if there is a way to "save" some of the cells which are assigned to negative, as I think this is most likely due to the high cutoff for HASH2 (?) 

If I reduce the positive.quantile of the HTOdemux function, the number of cells assigned to negative decreases but the number of doublet strongly increases and the number of cells assigned to HASH1 and 3 also drops, so this does not seem to be a solution.
Any idea?

One thing that I tried is to subset the "negative" cells and re-run HTOdemux on this fraction only to try to get out more assigned cells. However this seems to create unspecific signal, so also not a real option apparently.

Thank you in advance

**Comment by felipelopesassis** [comment]
Dear all, 

I'm having the same issue. I have a similar dataset (10xGenomics 5' Gene expression data + hashtags),  with ~4000 single-cells  stained with 2 #HTOs. HTO7 signal should also be 10X higher than HTO8. Below, I'm showing a snapshot of the CLR-normalized count data:  

### 
```
s@assays[["HTO"]]@counts
2 x 4073 sparse Matrix of class "dgCMatrix"
   [[ suppressing 71 column names ‘AAACCTGCAAACCCAT-1’, ‘AAACCTGCAAAGGCGT-1’, ‘AAACCTGCAAGACGTG-1’ ... ]]
                                                                                                                                                        
HTO7-PROT 3.21877165 3.27287164 3.41891149 3.08501699 3.29377209 3.23383731 3.24836495 3.20417320 3.21625773 
HTO8-PROT 0.03831886 0.03717339 0.03077135 0.04374802 0.03649156 0.03716288 0.03756261 0.03950117 0.03959125 
                                                                                                                                                        
HTO7-PROT 3.15643761 3.30775951 3.25414123 3.09127431 3.12638776 3.15436505 3.25094049 3.09960099 3.26432062 
HTO8-PROT 0.04203807 0.03524652 0.03820661 0.04596662 0.04295435 0.04234432 0.03722324 0.04511011 0.03348975
``` 

However, after running HTODemux most of the cells were labeled as negative. I played around with quantile, but nothing really helped. I also got the imbalanced cutoff threshold issue reported before. I hope someone can help us!

Thank you in advance

**Comment by ghost** [comment]
Hi, Sorry I'm late to the party. @felipelopesassis @erosix 

I had a similar issue. The cutoffs should be different since some HTOs are stronger than others; it's likely that more of that HTO was incubated with the sample or that it was incubated with fewer cells (or both) compared to the other HTOs. So using a quantile range is appropriate. It can be modified by positive.quantile, but sometimes you end up playing a game of maximizing the singlets vs anything else. Based on the standard figures in the seurat vignette (https://satijalab.org/seurat/articles/hashing_vignette.html); IF you are confident that it's wrong, I modified the HTOdemux function below so you can specify the positive.quantile_low (anything below is negative) or positive.quantile_high (if more than one 'high' CMOs it's a doublet), anything else is a singlet. Obviously use with caution and don't deviate too much from the default 0.99.

Code snippet:
`
library(cluster) #dependency
library(fitdistrplus) #dependency

MaxN <- function(x, N = 2){
  len <- length(x)
  if (N > len) {
    warning('N greater than length(x).  Setting N=length(x)')
    N <- length(x)
  }
  sort(x, partial = len - N + 1)[len - N + 1]
} #another dependency

HTODemux_low_highfun <- function (object, assay = "HTO", positive.quantile_low = 0.99, positive.quantile_high = 0.99, init = NULL, 
  nstarts = 100, kfunc = "clara", nsamples = 100, seed = 42, 
  verbose = TRUE) 
{
  if (!is.null(x = seed)) {
    set.seed(seed = seed)
  }
  assay <- assay %||% DefaultAssay(object = object)
  data <- GetAssayData(object = object, assay = assay)
  counts <- GetAssayData(object = object, assay = assay, slot = "counts")[, 
    colnames(x = object)]
  counts <- as.matrix(x = counts)
  ncenters <- init %||% (nrow(x = data) + 1)
  switch(EXPR = kfunc, kmeans = {
    init.clusters <- kmeans(x = t(x = GetAssayData(object = object, 
      assay = assay)), centers = ncenters, nstart = nstarts)
    Idents(object = object, cells = names(x = init.clusters$cluster)) <- init.clusters$cluster
  }, clara = {
    init.clusters <- clara(x = t(x = GetAssayData(object = object, 
      assay = assay)), k = ncenters, samples = nsamples)
    Idents(object = object, cells = names(x = init.clusters$clustering), 
      drop = TRUE) <- init.clusters$clustering
  }, stop("Unknown k-means function ", kfunc, ", please choose from 'kmeans' or 'clara'"))
  average.expression <- AverageExpression(object = object, 
    assays = assay, verbose = FALSE)[[assay]]
  if (sum(average.expression == 0) > 0) {
    stop("Cells with zero counts exist as a cluster.")
  }
  discrete_weak <- GetAssayData(object = object, assay = assay)
  discrete_weak[discrete_weak > 0] <- 0
  discrete_strong <- discrete_weak
  
  ###start modified section
  for (iter in rownames(x = data)) {
    values <- counts[iter, colnames(object)]
    values.use <- values[WhichCells(object = object, idents = levels(x = Idents(object = object))[[which.min(x = average.expression[iter, 
      ])]])]
    fit <- suppressWarnings(expr = fitdist(data = values.use, 
      distr = "nbinom"))
    cutoff <- as.numeric(x = quantile(x = fit, probs = positive.quantile_low)$quantiles[1])
    discrete_weak[iter, names(x = which(x = values > cutoff))] <- 1
    if (verbose) {
      message(paste0("Cutoff for ", iter, " : ", cutoff, 
        " reads"))
    }
  }
  
  for (iter in rownames(x = data)) {
  values <- counts[iter, colnames(object)]
  values.use <- values[WhichCells(object = object, idents = levels(x = Idents(object = object))[[which.min(x = average.expression[iter, 
    ])]])]
  fit <- suppressWarnings(expr = fitdist(data = values.use, 
    distr = "nbinom"))
  cutoff <- as.numeric(x = quantile(x = fit, probs = positive.quantile_high)$quantiles[1])
  discrete_strong[iter, names(x = which(x = values > cutoff))] <- 1
  if (verbose) {
    message(paste0("Cutoff for ", iter, " : ", cutoff, 
        " reads"))
    }
  }
  
  npositive_weak <- colSums(x = discrete_weak)
  npositive_weak[npositive_weak >= 1] <- "Singlet"        
  npositive_weak[npositive_weak == 0] <- "Negative"

  npositive_strong <- colSums(x = discrete_strong)
  npositive_strong[npositive_strong >= 2] <- "Doublet"
  npositive_strong[npositive_strong == 1 | npositive_strong == 0] <- "Singlet"
  
  
  classification.global <- data.frame(npositive_weak, npositive_strong)
  classification.global$npositive_strong[classification.global$npositive_weak == "Negative"] <- "Negative"
  classification.global <- classification.global$npositive_strong
  ###End modififed section
  
  donor.id = rownames(x = data)
  hash.max <- apply(X = data, MARGIN = 2, FUN = max)
  hash.maxID <- apply(X = data, MARGIN = 2, FUN = which.max)
  hash.second <- apply(X = data, MARGIN = 2, FUN = MaxN, N = 2)
  hash.maxID <- as.character(x = donor.id[sapply(X = 1:ncol(x = data), 
    FUN = function(x) {
      return(which(x = data[, x] == hash.max[x])[1])
    })])
  hash.secondID <- as.character(x = donor.id[sapply(X = 1:ncol(x = data), 
    FUN = function(x) {
      return(which(x = data[, x] == hash.second[x])[1])
    })])
  hash.margin <- hash.max - hash.second
  doublet_id <- sapply(X = 1:length(x = hash.maxID), FUN = function(x) {
    return(paste(sort(x = c(hash.maxID[x], hash.secondID[x])), 
      collapse = "_"))
  })
  classification <- classification.global
  classification[classification.global == "Negative"] <- "Negative"
  classification[classification.global == "Singlet"] <- hash.maxID[which(x = classification.global == 
    "Singlet")]
  classification[classification.global == "Doublet"] <- doublet_id[which(x = classification.global == 
    "Doublet")]
  classification.metadata <- data.frame(hash.maxID, hash.secondID, 
    hash.margin, classification, classification.global)
  colnames(x = classification.metadata) <- paste(assay, c("maxID", 
    "secondID", "margin", "classification", "classification.global"), 
    sep = "_")
  object <- AddMetaData(object = object, metadata = classification.metadata)
  Idents(object) <- paste0(assay, "_classification")
  doublets <- rownames(x = object[[]])[which(object[[paste0(assay, 
    "_classification.global")]] == "Doublet")]
  Idents(object = object, cells = doublets) <- "Doublet"
  object$hash.ID <- Idents(object = object)
  return(object)
}
`
